provider/src/main/kotlin/org/gradle/kotlin/dsl/TypeOfExtensions.kt

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.kotlin.dsl
+
+import org.gradle.api.reflect.TypeOf
+
+
+/**
+ * Creates an instance of [TypeOf] for the given parameterized type.
+ *
+ * @param T the type
+ * @return the [TypeOf] that captures the generic type of the given parameterized type
+ */
+inline fun <reified T> typeOf() =
+    object : TypeOf<T>() {}
```

Comment:
Love this. Is this intended for consumers though. Just making sure.
Seems like this trick is pretty common in libraries that need to save the reflected type.
https://github.com/JLLeitschuh/kotlin-guiced/blob/master/core/src/main/kotlin/org/jlleitschuh/guice/Key.kt#L5

Curiosity though, how would this handle nullable types?
Eg:
`typeOf<String?>`?

Reply:
`TypeOf` is Gradle public API, so yes, intended for consumers.

Good point, it wouldn't handle Kotlin nullability. So it should be `<reified T : Any>` in order to limit usage to non-nullable type parameters only.

Reply:
After thinking about this a bit more, `typeOf<T>()` should not limit usage to non-nullable types only. `the<T>()` and `configure<T>()` do because they throw if the underlying object cannot be found.

Other consumers of `typeOf<T>()` may pass nullable types. It just happens that nullability is ignored by the underlying java implementations. And nothing stops one to write `typeOf<NonNullable<Nullable?>>()` anyway.

Reply:
Fair enough.